README.md
# Video Game System SDK - Pico

## Description

Video Game System SDK - Pico (vgssdk-pico) は、[東方VGS実機版](https://github.com/suzukiplan/tohovgs-pico)及びその互換ハードウェア向けのアプリケーションを効率的に開発することを目的としたシミュレータと API を提供する SDK です。

vgssdk-pico を用いて開発したアプリは、OS 依存の API を使用しない限り、次のシステム環境下で同一の動作をします:

- [東方VGS実機版](https://github.com/suzukiplan/tohovgs-pico) (RaspberryPi Pico + LCD + DAC)
- macOS (SDL2 で [東方VGS実機版](https://github.com/suzukiplan/tohovgs-pico) をシミュレート)
- Linux (SDL2 で [東方VGS実機版](https://github.com/suzukiplan/tohovgs-pico) をシミュレート)

vgssdk-pico を用いることで、パソコン（Linux or macOS）のみで効率的に[東方VGS実機版](https://github.com/suzukiplan/tohovgs-pico)向けアプリケーションの開発ができるようになります。

> スマホアプリを iOS のシミュレータ や Android のエミュレータで開発するようなイメージのものと思っていただければ大丈夫です。

## PIN assign

![./rp2040.png](./rp2040.png)

|RP2040 (SoC)|ILI9341 (LCD)|FT6336U (CTP)|UDA1334A (DAC)|Joypad|
|:-|:-|:-|:-|:-|
|4: `GPIO2`|-|-|-|`A button`|
|5: `GPIO3`|-|-|-|`B button`|
|6: `GPIO4`|-|`CTP_SDA`|-|-|
|7: `GPIO5`|-|`CTP_SCK`|-|-|
|9: `GPIO6`|-|-|-|`Down`|
|10: `GPIO7`|-|-|-|`Up`|
|11: `GPIO8`|-|-|-|`Left`|
|12: `GPIO9`|-|-|-|`Right`|
|13: `GND`|-|-|-|`GND`|
|14: `GPIO10`|-|-|-|`Start`|
|15: `GPIO11`|-|-|-|`Select`|
|17: `GPIO13`|-|-|`DIN`|-|
|18: `GND`|-|-|`GND`|-|
|19: `GPIO14`|-|-|`BCLK`|-|
|20: `GPIO15`|-|-|`WSEL`|-|
|21: `GPIO16`|`SDO (MISO)`|-|-|-|
|22: `GPIO17`|`LCD_CS`|-|-|-|
|24: `GPIO18`|`SCK`|-|-|-|
|25: `GPIO19`|`SDI (MOSI)`|-|-|-|
|26: `GPIO20`|-|`CTP_RST`|-|-|
|27: `GPIO21`|-|`CTP_INT`|-|-|
|29: `GPIO22`|`LCD_RST`|-|-|-|
|32: `GPIO27`|`LED`|-|-|-|
|34: `GPIO28`|`LCD_RS or LCD_DC`|-|-|-|
|36: `3.3V`|`VCC`|-|-|-|
|38: `GND`|`GND`|-|-|-|
|40: `VOUT (5V)`|-|-|`VIN`|-|

> `Joypad` の配線については [コチラの記事](https://note.com/suzukiplan/n/ncccafb305eae) を参照してください。

## How to build example app

[./example/README.md](./example/README.md) を参照してください。

## API outline

vgssdk-pico は C++（C++11以降）用の次のクラス群を提供します。

- `VGS` ... メインコントローラ（シングルトン）
- `VGS::GFX` ... 画面描画機能を提供するクラス
- `VGS::VDP` ... 画面描画機能を提供するクラス（ゲーム用）
- `VGS::BGM` ... VGS波形メモリ音源機能を提供するクラス
- `VGS::SoundEffect` ... 効果音再生機能を提供するクラス　
- `VGS::IO` ... 外部デバイス I/O 機能（タッチパネル+ゲームパッド）を提供するクラス

アプリケーションは次のように実装します

1. [vgssdk.h](src/vgssdk.h) を `#include`
2. `extern VGS vgs;` で VGS インスタンス（グローバル変数）を参照
3. `extern "C" void vgs_setup() { }` に起動時の初期化処理を記述
4. `extern "C" void vgs_loop() { }` にメインループ処理を記述

> [サンプリアプリ](example) の実装を参照してください。

なお、[東方VGS実機版](https://github.com/suzukiplan/tohovgs-pico)が対応する SoC (RaspberryPi Pico) はマルチコアコア（2 cores）の構成ですが、片方のコアは VGS の音声再生に専念させるため vgssdk-pico を用いるアプリはシングルコア（シングルスレッド & シングルタスク）で設計するものとします。

## How to Integrate

[src](./src) ディレクトリ以下のファイルを対象プロジェクトに組み込んで利用してください。

|File name|RP2040|Simulator|Description|
|:---------------------------------------|:-:|:-:|:-|
|[FT6336U.hpp](./src/FT6336U.hpp)        | o | - |タッチパネル（FT6336U）のドライバ|
|[lz4.c](./src/lz4.c)                    | o | o |BGM ファイルの解凍|
|[lz4.h](./src/lz4.h)                    | o | o |〃|
|[vgsdecv.hpp](./src/vgsdecv.hpp)        | o | o |BGM デコーダ|
|[vgssdk_pico.cpp](./src/vgssdk_pico.cpp)| o | - |VGSSDK の RP2040 依存処理|
|[vgssdk_sdl2.cpp](./src/vgssdk_sdl2.cpp)| - | o |VGSSDK の SDL2 (macOS, Linux) 依存処理|
|[vgssdk.h](./src/vgssdk.h)              | o | o |VGSSDK のヘッダファイル|
|[vgstone.c](./src/vgstone.c)            | o | o |BGM の音色データ|

## Compile Flags

|Compile Flag|Description|
|:-|:-|
|`-DVGSBGM_LIMIT_SIZE=数値`|BGMの非圧縮サイズ上限を KB 単位で指定（省略時: `108` KB）|
|`-DVGSVDP_DISPLAY_LIMIT=数値`|VDPの表示領域サイズの上限を KB 単位で指定（省略時: `90` KB）|
|`-DVGSVDP_PTNTBL_ROM`|VDP::RAMのパターンテーブルを ROM 領域にする（32KB の RAM 領域を確保）<br>※若干性能が劣化する可能性があります|
|`-DVGSGFX_ROTATION=0`|画面の向きを Portrait にする|
|`-DVGSGFX_ROTATION=1`|画面の向きを Landscape にする|
|`-DVGSGFX_ROTATION=2`|画面の向きを Reverse Portrait にする __(省略時のデフォルト)__|
|`-DVGSGFX_ROTATION=3`|画面の向きを Reverse Landscape にする|

> __NOTE:__ RP2040 で使用できる RAM サイズは `264KB` のため、規定の状態で `VGS::BGM` と `VGS::VDP` (display: 90KB + VRAM: 40KB) の両方を使用すると `238KB` (90%以上) を専有してしまい、アプリケーションが利用できるメモリ残量が僅かになってしまうため、`VGSBGM_LIMIT_SIZE` でメモリサイズを調整することが事実上必須となります。（`VGSBGM_LIMIT_SIZE` の既定値は東方VGSが提供する全楽曲の非圧縮データサイズの中で最も大きなサイズを基準に設定されています）
>
> 【参考】`VGS::BGM` と `VGS::VDP` の両方を使用する [example/vdp](./example/vdp/vdp_test.cpp) を RP2040 向けに `-DVGSBGM_LIMIT_SIZE=32` を指定してビルドした時のメモリ専有量は 約190KB です。
>
> ```
> Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
> RAM:   [=======   ]  74.4% (used 195088 bytes from 262144 bytes)
> Flash: [==        ]  18.0% (used 375748 bytes from 2093056 bytes)
> ```

## `VGS class`

### Public Member Variables

- `vgs.gfx` ... [VGS::GFX class](#vgsgfx-class) の インスタンス
- `vgs.vdp` ... [VGS::VDP class](#vgsvdp-class) の インスタンス
- `vgs.bgm` ... [VGS::BGM class](#vgsbgm-class) の インスタンス
- `vgs.eff` ... [VGS::SoundEffect class](#vgssoundeffect-class) の インスタンス
- `vgs.io` ... [VGS::IO class](#vgsio-class) の インスタンス

### `VGS::delay method`

```c++
void VGS::delay(int ms);
```

指定した時間（ミリ秒）待機する

### `VGS::led method`

```c++
void VGS::led(bool on);
```

本体 LED の点灯・消灯

### `VGS::setFrameRate method`

```c++
void VGS::setFrameRate(int frameRate);
```

- シミュレータが動作するフレームレートを指定します
  - 0 〜 60 の範囲で指定できます
  - 0 を指定するとノーウェイト（デフォルト）になります
  - シミュレータがサポートするフレームレートに自動補正される場合があります
- [東方VGS実機版](https://github.com/suzukiplan/tohovgs-pico) では本メソッドを呼び出しても挙動は変化しません

### `VGS::getFrameRate method`

```c++
int VGS::getFrameRate();
```

- `VGS::setFrameRate` で指定したフレームレート（補正値）を取得します

## `VGS::GFX class`

### Constructor

`VGS::GFX class` は「物理ディスプレイ」向けと「仮想ディスプレイ」向け2種類のコンストラクタを提供しています。

```c++

CHANGELOG.md
# Change Log

## Version 1.5.0

Pull Request: https://github.com/suzukiplan/vgssdk-pico/pull/7

- ジョイパッドの入力をサポート
- bmp2img で 24 ビット Bitmap 形式をサポート

## Version 1.4.0

Pull Request: https://github.com/suzukiplan/vgssdk-pico/pull/6

- `VGS::VDP::resize` メソッドを追加
- vgsftv と vgslz4 コマンドの機能を vgsmml に統合

## Version 1.3.0

Pull Request: https://github.com/suzukiplan/vgssdk-pico/pull/5

- コンパイルオプション VGSVDP_PTNTBL_ROM の指定によりパターンテーブルに ROM 領域を指定できるようにする

## Version 1.2.1

Pull Request: https://github.com/suzukiplan/vgssdk-pico/pull/4

- `VGS::VDP::setOam` と `VGS::VDP::getOam` の引数 `index` の型を `unsigned char` から `int` に変更
- bugfix: VGSVDP_DISPLAY_LIMIT に KB 単位で指定とあるが実際は 指定サイズ x 2 KB となっていたため 指定サイズ KB となるように修正（既定値を45→90に変更）
- example のアセット類の管理構成を変更

## Version 1.2.0

Pull Request: https://github.com/suzukiplan/vgssdk-pico/pull/3

- VDPサポート
- 画面回転機能のサポート
  - 動的な回転はサポートしない方針
  - ビルド時にコンパイルフラグ VGSGFX_ROTATION で指定する仕様
- example/sound のアセットデータを tools で生成する形に変更

## Version 1.1.0

Pull Request: https://github.com/suzukiplan/vgssdk-pico/pull/2

- enhancement:
  - VDPサポート
  - BGM上限サイズ変更オプションを追加
  - 開発に役立つツール類を tohovgs-pico からコピー
  - 効果音機能の追加
  - example/sound の効果音対応
- bugfix:
  - `VGS::GFX::line` が実機で想定通りに動かない不具合を修正

## Version 1.0.0

- first stable



src/vgssdk_sdl2.cpp
/**
 * VGS SDK Pico for SDL2 (macOS and Linux)
 * License under MIT: https://github.com/suzukiplan/vgssdk-pico/blob/master/LICENSE.txt
 * (C)2023, SUZUKI PLAN
 */
#include "SDL.h"
#include "vgsdecv.hpp"
#include "vgssdk.h"
#include <chrono>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifndef VGSVDP_DISPLAY_LIMIT
#define VGSVDP_DISPLAY_LIMIT 90
#endif

#ifndef VGSGFX_ROTATION
#define VGSGFX_ROTATION 2 // default = reverse portrait
#endif

#if VGSGFX_ROTATION == 0
#define VGS_DISPLAY_WIDTH 240
#define VGS_DISPLAY_HEIGHT 320
#elif VGSGFX_ROTATION == 1
#define VGS_DISPLAY_WIDTH 320
#define VGS_DISPLAY_HEIGHT 240
#elif VGSGFX_ROTATION == 2
#define VGS_DISPLAY_WIDTH 240
#define VGS_DISPLAY_HEIGHT 320
#define REVERSE_SCREEN
#elif VGSGFX_ROTATION == 3
#define VGS_DISPLAY_WIDTH 320
#define VGS_DISPLAY_HEIGHT 240
#define REVERSE_SCREEN
#endif

#define abs_(x) (x >= 0 ? (x) : -(x))
#define sgn_(x) (x >= 0 ? (1) : (-1))

VGS vgs;
static SDL_Window* window;
static SDL_Surface* windowSurface;
static unsigned short vdp_display_buf[VGSVDP_DISPLAY_LIMIT * 512];
static VGS::VDP::RAM vdp_vram;
static SDL_AudioDeviceID bgmAudioDeviceId;
static bool bgmLoaded;

static void log(const char* format, ...)
{
    char buf[256];
    va_list args;
    va_start(args, format);
    vsnprintf(buf, sizeof(buf), format, args);
    va_end(args);
    puts(buf);
}

static inline unsigned int bit5to8(unsigned char bit5)
{
    unsigned char bit8 = bit5 | ((bit5 & 0b11100000) >> 5);
    unsigned int result = bit8;
    result <<= 8;
    result |= bit8;
    result <<= 8;
    result |= bit8;
    result <<= 8;
    result |= bit8;
    return result;
}

static inline unsigned int bit6to8(unsigned char bit6)
{
    unsigned char bit8 = bit6 | ((bit6 & 0b11000000) >> 6);
    unsigned int result = bit8;
    result <<= 8;
    result |= bit8;
    result <<= 8;
    result |= bit8;
    result <<= 8;
    result |= bit8;
    return result;
}

static inline unsigned int color16to32(unsigned short rgb565)
{
    unsigned int color32 = bit5to8((rgb565 & 0b1111100000000000) >> 8) & windowSurface->format->Rmask;
    color32 |= bit6to8((rgb565 & 0b0000011111100000) >> 3) & windowSurface->format->Gmask;
    color32 |= bit5to8((rgb565 & 0b0000000000011111) << 3) & windowSurface->format->Bmask;
    color32 |= windowSurface->format->Amask;
    return color32;
}

static inline unsigned short flip(unsigned short value)
{
    unsigned short result = (value & 0x00FF) << 8;
    result |= (value & 0xFF00) >> 8;
    return result;
}

VGS::GFX::GFX()
{
    log("create GFX for Physical Display (%d, %d)", VGS_DISPLAY_WIDTH, VGS_DISPLAY_HEIGHT);
    memset(&this->vDisplay, 0, sizeof(this->vDisplay));
    this->counter = 0;
    this->clearViewport();
}

VGS::GFX::GFX(int width, int height)
{
    log("create GFX for Virtual Display (%d, %d) = %dKB", width, height, width * height * 2 / 1024 + (width * height * 2 % 1024 ? 1 : 0));
    this->counter = 0;
    this->vDisplay.width = width;
    this->vDisplay.height = height;
    this->vDisplay.buffer = malloc(width * height * 2);
    memset(this->vDisplay.buffer, 0, width * height * 2);
    this->clearViewport();
}

VGS::GFX::~GFX()
{
    if (this->vDisplay.buffer) {
        free(this->vDisplay.buffer);
    }
}

void VGS::GFX::startWrite()
{
    this->counter++;
}

void VGS::GFX::endWrite()
{
    if (0 < this->counter) {
        this->counter--;
        if (0 == this->counter && !this->isVirtual()) {
            SDL_UpdateWindowSurface(window);
        }
    }
}

void VGS::GFX::startWriteSimulatorOnly()
{
    this->startWrite();
}

void VGS::GFX::endWriteSimulatorOnly()
{
    this->endWrite();
}

int VGS::GFX::getWidth()
{
    if (this->isVirtual()) {
        return this->vDisplay.width;
    } else {
        return VGS_DISPLAY_WIDTH;
    }
}

int VGS::GFX::getHeight()
{
    if (this->isVirtual()) {
        return this->vDisplay.height;
    } else {
        return VGS_DISPLAY_HEIGHT;
    }
}

void VGS::GFX::setViewport(int x, int y, int width, int height)
{
    this->viewport.x = x;
    this->viewport.y = y;
    this->viewport.width = width;
    this->viewport.height = height;
}

void VGS::GFX::clearViewport()
{
    this->viewport.x = 0;
    this->viewport.y = 0;
    if (this->isVirtual()) {
        this->viewport.width = this->vDisplay.width;
        this->viewport.height = this->vDisplay.height;
    } else {
        this->viewport.width = VGS_DISPLAY_WIDTH;
        this->viewport.height = VGS_DISPLAY_HEIGHT;
    }
}

void VGS::GFX::clear(unsigned short color)
{
    if (this->isVirtual()) {
        int ptr = 0;
        auto buffer = this->getVirtualBuffer();
        for (int y = 0; y < this->vDisplay.height; y++) {
            for (int x = 0; x < this->vDisplay.width; x++) {
                buffer[ptr++] = color;
            }
        }
    } else {
        auto color32 = color16to32(color);
        auto display = (unsigned int*)windowSurface->pixels;
        for (int y = 0; y < windowSurface->h; y++) {
            for (int x = 0; x < windowSurface->w; x++) {
                display[x] = color32;
            }
            display += windowSurface->pitch / windowSurface->format->BytesPerPixel;
        }
    }
}

void VGS::GFX::pixel(int x, int y, unsigned short color)
{
    if (this->viewport.width <= x || this->viewport.height <= y) {
        return;
    }
    x += this->viewport.x;
    y += this->viewport.y;
    if (x < 0 || y < 0) {
        return;
    }
    if (this->isVirtual()) {
        if (this->vDisplay.width <= x || this->vDisplay.height <= y) {
            return;
        }
        this->getVirtualBuffer()[y * this->vDisplay.width + x] = color;
    } else {
        if (VGS_DISPLAY_WIDTH <= x || VGS_DISPLAY_HEIGHT <= y) {
            return;
        }
        auto display = (unsigned int*)windowSurface->pixels;
        display += y * windowSurface->pitch / windowSurface->format->BytesPerPixel;
        display += x;
        *display = color16to32(color);
    }
}

void VGS::GFX::lineV(int x, int y, int height, unsigned short color)
{
    for (int i = 0; i < height; i++) {
        this->pixel(x, y++, color);
    }
}

void VGS::GFX::lineH(int x, int y, int width, unsigned short color)
{
    for (int i = 0; i < width; i++) {
        this->pixel(x++, y, color);
    }
}

void VGS::GFX::line(int x1, int y1, int x2, int y2, unsigned short color)
{
    if (x1 == x2) {
        this->lineV(x1, y1 < y2 ? y1 : y2, abs(y1 - y2) + 1, color);
    } else if (y1 == y2) {
        this->lineH(x1 < x2 ? x1 : x2, y1, abs(x1 - x2) + 1, color);
    }
    int ia, ib, ie;
    int w;
    int idx = x2 - x1;
    int idy = y2 - y1;
    w = 1;
    ia = abs_(idx);
    ib = abs_(idy);
    if (ia >= ib) {
        ie = -abs_(idy);
        while (w) {
            this->pixel(x1, y1, color);
            if (x1 == x2) {
                break;
            }
            x1 += sgn_(idx);
            ie += 2 * ib;
            if (ie >= 0) {
                y1 += sgn_(idy);
                ie -= 2 * ia;
            }
        }
    } else {
        ie = -abs_(idx);
        while (w) {
            this->pixel(x1, y1, color);
            if (y1 == y2) {
                break;
            }
            y1 += sgn_(idy);
            ie += 2 * ia;
            if (ie >= 0) {
                x1 += sgn_(idx);
                ie -= 2 * ib;
            }
        }
    }
}

void VGS::GFX::box(int x, int y, int width, int height, unsigned short color)
{
    if (0 < width && 0 < height) {
        this->lineH(x, y, width, color);
        this->lineH(x, y + height - 1, width, color);
        this->lineV(x, y, height, color);
        this->lineV(x + width - 1, y, height, color);
    }
}

void VGS::GFX::boxf(int x, int y, int width, int height, unsigned short color)
{
    if (0 < width && 0 < height) {
        for (int i = 0; i < height; i++) {
            this->lineH(x, y++, width, color);
        }
    }
}

void VGS::GFX::image(int x, int y, int width, int height, const unsigned short* buffer)
{
    int ptr = 0;
    for (int yy = 0; yy < height; yy++) {
        for (int xx = 0; xx < width; xx++) {
            this->pixel(x + xx, y + yy, flip(buffer[ptr++]));
        }
    }
}

void VGS::GFX::image(int x, int y, int width, int height, const unsigned short* buffer, unsigned short transparent)
{
    int ptr = 0;
    for (int yy = 0; yy < height; yy++) {
        for (int xx = 0; xx < width; xx++) {
            auto color = buffer[ptr++];
            if (color != transparent) {
                this->pixel(x + xx, y + yy, flip(color));
            }
        }
    }
}

void VGS::GFX::push(int x, int y)
{
    if (!this->isVirtual()) return;
    int ptr = 0;
    for (int yy = 0; yy < this->getHeight(); yy++) {
        for (int xx = 0; xx < this->getWidth(); xx++) {
            vgs.gfx.pixel(x + xx, y + yy, this->getVirtualBuffer()[ptr++]);
        }
    }
}

bool VGS::VDP::create(int width, int height)
{
    if (!this->resize(width, height)) {
        return false;
    }
    memset(this->display.buf, 0, width * height * 2);
    memset(this->vram, 0, sizeof(VDP::RAM));
    return true;
}

bool VGS::VDP::resize(int width, int height)
{
    if (sizeof(vdp_display_buf) < width * height * 2) {
        return false;
    }
    this->display.width = width;
    this->display.height = height;
    this->display.buf = vdp_display_buf;
    this->vram = &vdp_vram;
    return true;
}

void VGS::VDP::render(int x, int y)
{
    this->bgToDisplay();
    this->spriteToDisplay();
    vgs.gfx.image(x, y, this->display.width, this->display.height, this->display.buf);
}

static inline void playSoundEffect(VGS::SoundEffect* eff, short* buffer, int count)
{
    if (!eff->context.ptr || eff->context.masterVolume < 1) {
        return;
    }
    for (int i = 0; i < count; i++) {
        int wav = buffer[i];
        int snd = eff->context.ptr[eff->context.cursor];
        snd *= eff->context.masterVolume;
        snd /= 100;
        wav += snd;
        if (32767 < wav) {
            wav = 32767;
        } else if (wav < -32768) {
            wav = -32768;
        }
        buffer[i] = (short)wav;
        eff->context.cursor++;
        if (eff->context.count <= eff->context.cursor) {
            eff->context.ptr = nullptr;
            return;
        }
    }
}

static void bgmCallback(void* userdata, Uint8* stream, int len)
{
    auto eff = &((VGS*)userdata)->eff;
    if (bgmLoaded) {
        auto bgm = &((VGS*)userdata)->bgm;
        auto decoder = (VGSDecoder*)bgm->getContext();
        decoder->execute(stream, len);
        playSoundEffect(eff, (short*)stream, len / 2);
    } else {
        memset(stream, 0, len);
        playSoundEffect(eff, (short*)stream, len / 2);
    }
}

VGS::BGM::BGM()
{
    this->context = new VGSDecoder();
    bgmLoaded = false;
}

VGS::BGM::~BGM()
{
    delete (VGSDecoder*)this->context;
}

void VGS::BGM::pause()
{
    if (bgmAudioDeviceId && !this->paused) {
        SDL_PauseAudioDevice(bgmAudioDeviceId, 1);
        this->paused = true;
    }
}

void VGS::BGM::resume()
{
    if (bgmAudioDeviceId && this->paused) {
        SDL_PauseAudioDevice(bgmAudioDeviceId, 0);
        this->paused = false;
    }
}

void VGS::BGM::load(const void* buffer, size_t size)
{
    if (bgmAudioDeviceId) {
        this->pause();
        SDL_LockAudioDevice(bgmAudioDeviceId);
        bgmLoaded = ((VGSDecoder*)this->context)->load(buffer, size);
        SDL_UnlockAudioDevice(bgmAudioDeviceId);
        this->resume();
    }
}

int VGS::BGM::getMasterVolume()
{
    return ((VGSDecoder*)this->context)->getMasterVolume();
}

void VGS::BGM::setMasterVolume(int masterVolume)
{
    ((VGSDecoder*)this->context)->setMasterVolume(masterVolume);
}

void VGS::BGM::fadeout()
{
    ((VGSDecoder*)this->context)->fadeout();
}

void VGS::BGM::seekTo(int time, void (*callback)(int percent))
{
    if (bgmAudioDeviceId) {
        this->pause();
        ((VGSDecoder*)this->context)->seekTo(time, callback);
        this->resume();
    }
}

bool VGS::BGM::isPlayEnd()
{
    return ((VGSDecoder*)this->context)->isPlayEnd();
}

int VGS::BGM::getLoopCount()
{
    return ((VGSDecoder*)this->context)->getLoopCount();
}

unsigned char VGS::BGM::getTone(int cn)
{
    if (0 <= cn && cn < 6) {
        return ((VGSDecoder*)this->context)->getTone(cn & 0xFF);
    } else {
        return 0xFF;
    }
}

unsigned char VGS::BGM::getKey(int cn)
{
    if (0 <= cn && cn < 6) {
        return ((VGSDecoder*)this->context)->getKey(cn & 0xFF);
    } else {
        return 0xFF;
    }
}

unsigned int VGS::BGM::getLengthTime()
{
    return ((VGSDecoder*)this->context)->getLengthTime();
}

unsigned int VGS::BGM::getLoopTime()
{
    return ((VGSDecoder*)this->context)->getLoopTime();
}

unsigned int VGS::BGM::getDurationTime()
{
    return ((VGSDecoder*)this->context)->getDurationTime();
}

int VGS::BGM::getIndex()
{
    return ((VGSDecoder*)this->context)->getIndex();
}

VGS::SoundEffect::SoundEffect()
{
    memset(&this->context, 0, sizeof(this->context));
    this->context.masterVolume = 100;
}

void VGS::SoundEffect::play(const short* buffer, size_t size)
{
    SDL_LockAudioDevice(bgmAudioDeviceId);
    this->context.ptr = buffer;
    this->context.count = size / 2;
    this->context.cursor = 0;
    SDL_UnlockAudioDevice(bgmAudioDeviceId);
}

VGS::VGS()
{
    this->halt = false;
    this->frameRate = 0;
}

VGS::~VGS()
{
}

void VGS::delay(int ms)
{
    usleep(ms * 1000);
}

void VGS::led(bool on)
{
    log("LED: %s", on ? "on" : "off");
}

void VGS::setFrameRate(int frameRate)
{
    if (frameRate < 1) {
        this->frameRate = 0;
    } else if (frameRate < 15) {
        this->frameRate = 10;
    } else if (frameRate < 25) {
        this->frameRate = 20;
    } else if (frameRate < 35) {
        this->frameRate = 30;
    } else if (frameRate < 45) {
        this->frameRate = 40;
    } else if (frameRate < 55) {
        this->frameRate = 50;
    } else {
        this->frameRate = 60;
    }
}

int main()
{
    log("Booting VGS for SDL2.");
    SDL_version sdlVersion;
    SDL_GetVersion(&sdlVersion);
    log("SDL version: %d.%d.%d", sdlVersion.major, sdlVersion.minor, sdlVersion.patch);

    log("init SDL");
    if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_EVENTS)) {
        log("SDL_Init failed: %s", SDL_GetError());
        exit(-1);
    }

    log("init AudioDriver");
    SDL_AudioSpec desired;
    SDL_AudioSpec obtained;
    desired.freq = 22050;
    desired.format = AUDIO_S16LSB;
    desired.channels = 1;
    desired.samples = 2048;
    desired.callback = bgmCallback;
    desired.userdata = &vgs;
    bgmAudioDeviceId = SDL_OpenAudioDevice(nullptr, 0, &desired, &obtained, 0);
    if (0 == bgmAudioDeviceId) {
        log(" ... SDL_OpenAudioDevice failed: %s", SDL_GetError());
        exit(-1);
    }
    log("- obtained.freq = %d", obtained.freq);
    log("- obtained.format = %X", obtained.format);
    log("- obtained.channels = %d", obtained.channels);
    log("- obtained.samples = %d", obtained.samples);
    SDL_PauseAudioDevice(bgmAudioDeviceId, 0);

    log("create SDL window");
    window = SDL_CreateWindow(
        "VGS for SDL2",
        SDL_WINDOWPOS_UNDEFINED,
        SDL_WINDOWPOS_UNDEFINED,
        VGS_DISPLAY_WIDTH,
        VGS_DISPLAY_HEIGHT,
        SDL_WINDOW_OPENGL);

    log("Get SDL window surface");
    windowSurface = SDL_GetWindowSurface(window);
    if (!windowSurface) {
        log("SDL_GetWindowSurface failed: %s", SDL_GetError());
        exit(-1);
    }

    log("PixelFormat: %d bits (%d bytes)", (int)windowSurface->format->BitsPerPixel, (int)windowSurface->format->BytesPerPixel);
    log("Rmask: %08X", (int)windowSurface->format->Rmask);
    log("Gmask: %08X", (int)windowSurface->format->Gmask);
    log("Bmask: %08X", (int)windowSurface->format->Bmask);
    log("Amask: %08X", (int)windowSurface->format->Amask);

    if (4 != windowSurface->format->BytesPerPixel) {
        log("unsupported pixel format (support only 4 bytes / pixel)");
        exit(-1);
    }

    log("VDP info");
    log("- display limit: %ld bytes (%ldKB)", sizeof(vdp_display_buf), sizeof(vdp_display_buf) / 1024);
    log("- VRAM size: %ld bytes (%ldKB)", sizeof(vdp_vram), sizeof(vdp_vram) / 1024 + (sizeof(vdp_vram) % 1024 ? 1 : 0));

    vgs_setup();
    SDL_UpdateWindowSurface(window);

    log("Continue to execute vgs_loop while no stop signal is detected...");
    SDL_Event event;
    unsigned int loopCount = 0;
    const int waitFps0[3] = {0, 0, 0};
    const int waitFps10[3] = {100, 100, 100};
    const int waitFps20[3] = {50, 50, 50};
    const int waitFps30[3] = {34, 33, 33};
    const int waitFps40[3] = {25, 25, 25};
    const int waitFps50[3] = {20, 20, 20};
    const int waitFps60[3] = {17, 17, 16};
    const int* waitTime[7] = {waitFps0, waitFps10, waitFps20, waitFps30, waitFps40, waitFps50, waitFps60};

    while (!vgs.halt) {
        loopCount++;
        auto start = std::chrono::system_clock::now();
        if (0 < vgs.getFrameRate()) {
            vgs.gfx.startWrite();
        }
        vgs_loop();
        bool quit = false;
        while (SDL_PollEvent(&event)) {
            if (event.type == SDL_QUIT) {
                quit = true;
                break;
            } else if (event.type == SDL_KEYDOWN) {
                switch (event.key.keysym.sym) {
                    case SDLK_q: quit = true; break;
                    case SDLK_UP: vgs.io.joypad.up = true && vgs.io.isJoypadEnabled(); break;
                    case SDLK_DOWN: vgs.io.joypad.down = true && vgs.io.isJoypadEnabled(); break;
                    case SDLK_LEFT: vgs.io.joypad.left = true && vgs.io.isJoypadEnabled(); break;
                    case SDLK_RIGHT: vgs.io.joypad.right = true && vgs.io.isJoypadEnabled(); break;
                    case SDLK_z: vgs.io.joypad.b = true && vgs.io.isJoypadEnabled(); break;
                    case SDLK_x: vgs.io.joypad.a = true && vgs.io.isJoypadEnabled(); break;
                    case SDLK_SPACE: vgs.io.joypad.start = true && vgs.io.isJoypadEnabled(); break;
                    case SDLK_ESCAPE: vgs.io.joypad.select = true && vgs.io.isJoypadEnabled(); break;
                }
                if (quit) {
                    break;
                }
            } else if (event.type == SDL_KEYUP) {
                switch (event.key.keysym.sym) {
                    case SDLK_UP: vgs.io.joypad.up = false; break;
                    case SDLK_DOWN: vgs.io.joypad.down = false; break;
                    case SDLK_LEFT: vgs.io.joypad.left = false; break;
                    case SDLK_RIGHT: vgs.io.joypad.right = false; break;
                    case SDLK_z: vgs.io.joypad.b = false; break;
                    case SDLK_x: vgs.io.joypad.a = false; break;
                    case SDLK_SPACE: vgs.io.joypad.start = false; break;
                    case SDLK_ESCAPE: vgs.io.joypad.select = false; break;
                }
            } else if (event.type == SDL_MOUSEBUTTONDOWN) {
                vgs.io.touch.on = true;
                SDL_GetMouseState(&vgs.io.touch.x, &vgs.io.touch.y);
            } else if (event.type == SDL_MOUSEBUTTONUP) {
                vgs.io.touch.on = false;
            } else {
                SDL_GetMouseState(&vgs.io.touch.x, &vgs.io.touch.y);
            }
        }
        if (quit) {
            break;
        }
        if (0 < vgs.getFrameRate()) {
            vgs.gfx.endWrite();
            std::chrono::duration<double> diff = std::chrono::system_clock::now() - start;
            int ms = (int)(diff.count() * 1000);
            int wait = waitTime[vgs.getFrameRate() / 10][loopCount % 3];
            if (ms < wait) {
                usleep((wait - ms) * 1000);
            }
        }
    }

    if (vgs.halt) {
        log("halted by app");
    } else {
        log("halted by system");
    }

    vgs.bgm.pause();
    SDL_CloseAudioDevice(bgmAudioDeviceId);
    bgmAudioDeviceId = 0;
    SDL_DestroyWindow(window);
    SDL_Quit();
    return 0;
}
